src/VisualStudio/Core/Def/Packaging/PackageSearchService.cs

```diff
@@ -0,0 +1,614 @@
+using Elfie.Model.Structures;
+using Elfie.Model.Tree;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Settings;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell.Settings;
+using static System.FormattableString;
+using VSShell = Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Packaging
+{
+    internal partial class PackageSearchService : ForegroundThreadAffinitizedObject, IPackageSearchService, IDisposable
+    {
+        // Internal for testing purposes.
+        internal const string ContentAttributeName = "content";
+        internal const string UpToDateAttributeName = "upToDate";
+        internal const string TooOldAttributeName = "tooOld";
+
+        private const string HostId = "RoslynNuGetSearch";
+        private static readonly LinkedList<string> s_log = new LinkedList<string>();
+
+        private readonly int _dataFormatVersion = AddReferenceDatabase.TextFileFormatVersion;
+
+        private readonly object _gate = new object();
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly CancellationToken _cancellationToken;
+
+        private AddReferenceDatabase _database_doNotAccessDirectly;
+
+        private readonly DirectoryInfo _cacheDirectoryInfo;
+        private readonly FileInfo _databaseFileInfo;
+
+        private readonly IPackageSearchDelayService _delayService;
+        private readonly IPackageSearchIOService _ioService;
+        private readonly IPackageSearchLogService _logService;
+        private readonly IPackageSearchRemoteControlService _remoteControlService;
+        private readonly IPackageSearchPatchService _patchService;
+        private readonly IPackageSearchDatabaseFactoryService _databaseFactoryService;
+        private readonly Func<Exception, bool> _swallowException;
+
+        public PackageSearchService(VSShell.SVsServiceProvider serviceProvider)
+            : this(CreateRemoteControlService(serviceProvider),
+                   new LogService((IVsActivityLog)serviceProvider.GetService(typeof(SVsActivityLog))),
+                   new DelayService(),
+                   new IOService(),
+                   new PatchService(),
+                   new DatabaseFactoryService(),
+                   new ShellSettingsManager(serviceProvider).GetApplicationDataFolder(ApplicationDataFolder.LocalSettings),
+                   // swallow all exceptions
+                   e => true,
+                   new CancellationTokenSource())
+        {
+            // Kick off a database update.  Wait a few seconds before starting so we don't
+            // interfere too much with solution loading.
+            Task.Delay(TimeSpan.FromSeconds(10)).ContinueWith(_ => UpdateDatabaseInBackgroundAsync(), TaskScheduler.Default);
```

Comment:
How does the diagnostic engine know when solution loading has completed before starting to issue diagnostics? I'd expect this to work like that system (or any other systems that know to wait for solution load).